basic API - book store managment

Avigailbr opened the pull request.

Avigailbr committed:
basic API - book store managment

jonathanAlu commented on models/book.go:
```diff
@@ -0,0 +1,48 @@
+type UpdateBook struct {
+	Id          string     `json:"id" binding:"required`
+	Title       string     `json:"title" binding:"required"`
+}
```
put the structs on the top of your files

jonathanAlu replied:
I'm not sure that the model is the place to put the `UpdateBook` struct.
If it's only for the client request process you can put it there or even use anonymous struct.. something like: 
```
var updateBook struct {
    Id          string     `json:"id" binding:"required`
    Title       string     `json:"title" binding:"required"`
}
```

jonathanAlu commented on cmd/service/main.go:
```diff
@@ -0,0 +1,30 @@
+		fmt.Printf("Initialization failed, %v" ,err.Error())
+		return
+	}
+	controllers.BookController = controllers.NewElasticBookController()
```
why do you initialise it here if you are not using it?

jonathanAlu commented on cmd/service/main.go:
```diff
@@ -0,0 +1,30 @@
+	router := gin.New()
+
+	// middelwares
+	router.Use(gin.Logger(), gin.Recovery(), middlewares.WriteActionToRedis())
```
❤️ the use of the redis middleware

jonathanAlu commented on cmd/service/main.go:
```diff
@@ -0,0 +1,30 @@
+	service.Routes(router)
+
+	router.Run(":8888")
+	//panic(connectors.BookConnectors.Shutdown) - TODO (ASK)- do i need to run `router.Run` inside a goroutine
```
no.. only if you want to do something after the server starts, for example, waiting for kill interrupt..

jonathanAlu commented on config/config.go:
```diff
@@ -0,0 +1,13 @@
+	IndexName = "books"
+	TypeName = "book"
+	SearchResultFrom = 0
+	SearchResultSize = 10
```
i'm not sure that the elastic index and search params are part of the config package.  only the url IMO

jonathanAlu commented on connectors/redis.go:
```diff
@@ -0,0 +1,20 @@
+	client := redis.NewClient(&redis.Options{
+		Addr:     addr,
+		Password: "", // no password set
+		DB:       0,
```
what is the DB option?

Avigailbr replied:
which DB to use, 0 its for default (I will add a comment)

jonathanAlu commented on connectors/setup.go:
```diff
@@ -0,0 +1,33 @@
+	_, err = BookConnectors.NewRedisClient(config.RedisAdd, config.RedisDB)
+	if err != nil {
+		return err
+	}
```
i would initialise the connectors in the end and use the return values:

```suggestion
	esClient, err := BookConnectors.NewElasticClient(config.ElasticUrl, config.IndexName, config.TypeName)
	if err != nil {
		return err
	}
	redisClient, err = BookConnectors.NewRedisClient(config.RedisAdd, config.RedisDB)
	if err != nil {
		return err
	}
	
	BookConnectors = &Connectors{
	    ElasticClient: esClient,
 	    RedisClients: redisClient,
	}
```
Not critical but looks better IMO

jonathanAlu commented on connectors/elastic.go:
```diff
@@ -0,0 +1,26 @@
+		Index:   index,
+		DocType: doctype,
+	}
+	c.ElasticClient = elastic
```
I think it will be better to use the returned client instead of saving it directly from here.

Avigailbr replied:
I agree!

jonathanAlu commented on connectors/redis.go:
```diff
@@ -0,0 +1,20 @@
+		return nil, err
+	}
+
+	c.RedisClients = client
```
same as the elastic comment

jonathanAlu commented on connectors/setup.go:
```diff
@@ -0,0 +1,33 @@
+
+type Connectors struct {
+	ElasticClient *ElasticClient
+	RedisClients  *redis.Client
```
typo? RedisClient (singular)

Avigailbr replied:
yep! :)

jonathanAlu commented on connectors/setup.go:
```diff
@@ -0,0 +1,33 @@
+	return nil
+}
+
+func (c *Connectors) Shutdown() {
```
are you using it? maybe i've missed it..

Avigailbr replied:
Nope.. 
I thought that if the server would run inside a goroutine, I would need to managed the shutdown 
(I will delete it)

jonathanAlu commented on controllers/bookController.go:
```diff
@@ -0,0 +1,219 @@
+	BookController IBookController
+)
+
+type IBookController interface {
+	GetBook(id string) (*models.Book, error)
+	InsertBook(book *models.Book) (string, error)
+	UpdateBook(id string, fields map[string]interface{}) error
+	DeleteBook(id string) error
+	Search(fields map[string]string) ([]models.Book, error)
+	Info() (map[string]interface{}, error)
+}
```
I'm not sure we need an interface here.. let's talk about it

Avigailbr replied:
for sure! It's only if we decide to use a different DB  from elastic

jonathanAlu commented on controllers/bookController.go:
```diff
@@ -0,0 +1,219 @@
+	Info() (map[string]interface{}, error)
+}
+
+type ElasticBookController struct {
+}
+
+func NewElasticBookController() IBookController {
+	return &ElasticBookController{}
+}
```
I don't think that all of this methods should be methods..
I believe that functions is the way to go.
https://flaviocopes.com/golang-methods-or-functions/
If you choose to change to functions, you won't need this struct..

Avigailbr replied:
I got it!  (it's a good article)
I think it's all because the use of the interface

jonathanAlu commented on models/error.go:
```diff
@@ -0,0 +1,15 @@
+package models
+
+type StringError struct {
```
The basic go `error` class has a method `Error()` which returns a string so i guess you don't need this class

Avigailbr replied:
my bad!

jonathanAlu commented on models/book.go:
```diff
@@ -0,0 +1,48 @@
+const ctLayout = "2006-01-02T00:00:00Z"
+
+// TODO (ASK) - where to put this?
+type CustomTime struct {
```
why are you not just using `time.TIme`? you are basically only wrapping it without adding any new functionality, no?

Avigailbr replied:
it's for parsing the time from elastic to go, and vice versa

jonathanAlu commented on middlewares/writeaction.go:
```diff
@@ -0,0 +1,49 @@
+		}
+		c.Next()
+
+		if err := WriteToRedis(username, c.Request.Method, c.Request.RequestURI); err != nil {
```
You can use go routine and process the request in parallel.

Avigailbr replied:
to wrap the return func in a goroutine?

jonathanAlu commented on middlewares/writeaction.go:
```diff
@@ -0,0 +1,49 @@
+	value := "method: " + method + ", route: " + route
+
+	rdb := connectors.BookConnectors.RedisClients
+	_, err := rdb.ZAdd(userId, redis.Z{score, value}).Result()
```
you can use redis list, it saves the insertion order and than you don't need to use the global counter..
https://redis.io/commands/lpush

Avigailbr replied:
cool! it's working much better :)

jonathanAlu commented on models/book.go:
```diff
@@ -0,0 +1,48 @@
+	Title       string     `json:"title" binding:"required"`
+	AuthorName  string     `json:"author_name" binding:"required"`
+	Price       float64    `json:"price" binding:"required"`
+	Ebook       bool       `json:"ebook, ebook_available"`
```
ebook is not required? and also, are you using the 2 forms: `ebook, ebook_available`?

Avigailbr replied:
yes (it's in the exercise instruction)
the 2 forms is:
1. When Inserting a new book, the user JSON contains "ebook_available" field. 
2. inside the searching function, the result from the elastic is with "ebook" field"

jonathanAlu commented on service/handlers.go:
```diff
@@ -0,0 +1,146 @@
+func UpdateBookTitle(c *gin.Context) {
+
+	var book models.UpdateBook
+	if err := c.ShouldBind(&book); err != nil {
```
I think that it will be best to hold a business model obj book (if needed) and a simple DTO (https://en.wikipedia.org/wiki/Data_transfer_object) for the requests with the json annotation and eveything.
What do you think?

jonathanAlu commented on service/handlers.go:
```diff
@@ -0,0 +1,146 @@
+
+func activityInfo(c *gin.Context) {
+	username := c.Query("username")
+	rdb := connectors.BookConnectors.RedisClients
+
+	items, err := rdb.ZRevRangeWithScores(username, 0, 2).Result()
+	if err != nil {
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	var actions []string
+	for _, zItem := range items {
+		actions = append(actions, fmt.Sprint(zItem.Member))
+	}
```
you are doing all the other api's logic in the controller so I believe this logic should be there as well :)

Avigailbr replied:
the "activityInfo" function logic?

jonathanAlu commented on controllers/bookController.go:
```diff
@@ -0,0 +1,219 @@
+	} else if author, ok := fields["author_name"]; ok && author != "" {
+		authorMatchQuery := elastic.NewMatchQuery("author_name", author)
+		query.Must(authorMatchQuery)
+	}
```
you can just use an if for each param, instead of the nesting..
```suggestion
	if title, ok := fields["title"]; ok && title != "" {
		titleMatchQuery := elastic.NewMatchQuery("title", title)
		query.Must(titleMatchQuery)
	}
	if author, ok := fields["author_name"]; ok && author != "" {
		authorMatchQuery := elastic.NewMatchQuery("author_name", author)
		query.Must(authorMatchQuery)
	}
```

jonathanAlu commented on controllers/bookController.go:
```diff
@@ -0,0 +1,219 @@
+			return nil, models.NewStringError("Conversion failed for `Price_range` field")
+		}
+		priceFilterQuery := elastic.NewRangeQuery("price").From(from).To(to)
+		query.Filter(priceFilterQuery)
```
❤️ the use of filter here vs match above!

jonathanAlu commented on controllers/bookController.go:
```diff
@@ -0,0 +1,219 @@
+		Query(query).
+		From(config.SearchResultFrom).
+		Size(config.SearchResultSize).
+		Pretty(true).
```
the pretty is for debug purpose?

Avigailbr replied:
it's Unnecessary (it enables the caller to indent the JSON output, but I'm not using it..)

jonathanAlu commented on controllers/bookController.go:
```diff
@@ -0,0 +1,219 @@
+		if err != nil {
+			return nil, err
+		}
+		json.Unmarshal(data, &book)
```
```suggestion
		err = json.Unmarshal(data, &book)
```

Avigailbr replied:
You right! 
it's weird, I'm asking if err !=nil, but I didn't assign err to the result from json.Unmarshal function

jonathanAlu commented on controllers/bookController.go:
```diff
@@ -0,0 +1,219 @@
+	}
+
+	var authorsCount int
+	authors, found := sr.Aggregations.Terms("author_count")
+	if found {
+		for _, count := range authors.Aggregations {
+			if i, err := json.Marshal(&count); err != nil {
+				return nil, err
+			} else if authorsCount, err = strconv.Atoi(string(i)); err != nil {
+				return nil, err
+			}
+		}
+		info["author_count"] = authorsCount
+		info["books_count"] = sr.Hits.TotalHits
+
+	} else {
+		return nil, &elastic.Error{Status: http.StatusNotFound}
+	}
```
I think that the cardinality aggregation return the number of different authors straight away ( https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-metrics-cardinality-aggregation.html)
So maybe you can just use the agg result as is:
```suggestion
	authorsCount, found := sr.Aggregations["author_count"]
	if found {
		info["author_count"] = authorsCount
		info["books_count"] = sr.Hits.TotalHits
	} else {
		return nil, &elastic.Error{Status: http.StatusNotFound}
	}
```
Just make sure I'm not completely off here :)

Avigailbr replied:
its works! I have a question about this..

jonathanAlu reviewed (commented):
Great job for your first go proj!!

Avigailbr committed:
Update connectors/setup.go

Co-authored-by: Jonathan Alumot <53745844+jonathanAlu@users.noreply.github.com>

Avigailbr committed:
add DATABASE middelware and reorganized the code

Avigailbr committed:
add DATABASE middelware and reorganized the code

jonathanAlu commented on cmd/service/main.go:
```diff
@@ -0,0 +1,18 @@
+	router := gin.New()
+
+	// middelwares
+	router.Use(gin.Logger(), gin.Recovery(),middlewares.Database(), middlewares.WriteActionToCache())
```
Why did you choose to do the `middlewares.Database()` as a middleware? 
Why not just once?

Avigailbr replied:
I thought it was a nicer way to init the connectors. (from her and not from "main")
The init happened only once. and for each request its sets the KEYS of "DB" and "cache"

jonathanAlu commented on datastore/datastore.go:
```diff
@@ -0,0 +1,17 @@
+type IActivityCacher interface {
+	AddAction(userId, method, route string) error
+	GetLastActions(username string) ([]string, error)
+}
```
🤩

jonathanAlu commented on service/handlers.go:
```diff
@@ -0,0 +1,152 @@
+
+func BookInfo(c *gin.Context) {
+	id := c.Param("id")
+	db := c.MustGet("DB").(datastore.IBookStorer)
```
I'm not familiar with this syntax but if it works I guess it's ok.
you can always use a reg var as well :)

jonathanAlu reviewed (approved):
Looks great!!

UrielBiton commented on .idea/book_service.iml:
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Generally speaking try to avoid pushing these type o files to git (any IDE related files)

UrielBiton commented on datastore/datastore.go:
```diff
@@ -0,0 +1,17 @@
+
+import "github.com/avigailbr/book_service/models"
+
+type IBookStorer interface {
```
Golang's interfaces naming convention is the suffix `er`, which you followed, but omit the `I`.

You can read more best practices [here](https://golang.org/doc/effective_go.html)

UrielBiton commented on datastore/datastore.go:
```diff
@@ -0,0 +1,17 @@
+	InsertBook(book *models.Book) (string, error)
+	UpdateBook(id string, fields map[string]interface{}) error
+	DeleteBook(id string) error
+	Search(fields map[string]string) ([]models.Book, error)
```
Not saying you necessarily need to do it - but since interfaces are implicit, you could split the interfaces to multiple functionalities - eg. `BookStorer`, `BookSearcher`, etc...

UrielBiton commented on datastore/elastic.go:
```diff
@@ -0,0 +1,204 @@
+	return esBookStore, nil
+}
+
+func (e ElasticBookStore) GetBook(id string) (*models.Book, error) {
```
You could omit the `Book` suffix from all the methods - you're in the context of a book store, you're getting, inserting or deleting books.

UrielBiton commented on datastore/elastic.go:
```diff
@@ -0,0 +1,204 @@
+	return nil
+}
+
+func (e ElasticBookStore) Search(fields map[string]string) ([]models.Book, error) {
```
I think it's better to have something more structured for search (you're a book store :) ) but that's just my personal taste

UrielBiton commented on datastore/redis.go:
```diff
@@ -0,0 +1,49 @@
+	Client *connectors.RedisClient
+}
+
+func NewRedisCache(addr string) (IActivityCacher, error) {
```
That's a common practice in languages such C# and Java, but since [golang is a bit different](https://medium.com/@cep21/what-accept-interfaces-return-structs-means-in-go-2fe879e25ee8) it's worth taking into account

UrielBiton commented on service/routes.go:
```diff
@@ -0,0 +1,21 @@
+
+func Routes(router *gin.Engine) *gin.Engine {
+	router.GET("/ping", Ping)
+	router.GET("/book/:id", BookInfo)
+	router.PUT("/book", AddBook)
+	router.POST("/book", UpdateBookTitle)
+	router.DELETE("/book/:id", DeleteBook)
```
You can have a group of routes with a shared prefix together, see how we did it in `go-searcher`

UrielBiton commented on middlewares/writeaction.go:
```diff
@@ -0,0 +1,32 @@
+var counter float64
+
+
+func WriteActionToCache() gin.HandlerFunc {
```
Liked it a lot! :)
Totally an interceptor

UrielBiton commented on middlewares/writeaction.go:
```diff
@@ -0,0 +1,32 @@
+	"net/http"
+)
+
+var counter float64
```
Is it being used? if so - where? couldn't find it

[I don't like package scoped variables](https://dave.cheney.net/2017/06/11/go-without-package-scoped-variables), unless there's a really good reason to hold them

Avigailbr replied:
Nope, I forgot to delete it :|

UrielBiton reviewed (approved):
Great work @Avigailbr ! Looks good :)

Avigailbr committed:
refactor:
	1) added .gitignore file
	2) return a struct instance instead of an interface (redis + elastic clients)
